### PR TITLE
Add registers and improve AST

### PR DESCRIPTION
**Pull request by LucasPickering**
- Combined some common parts of the two ASTs by pulling most of the instructions out into a common `Operator` type
- Added registers and used them for all the instructions
- lit

### CODE REVIEW

File: api/src/lang/machine.rs [hunk 1]
```diff
@@ -117,10 +117,32 @@ impl MachineState {
...
         input.reverse(); // Reverse the vec so we can pop off the values in order
         Self {
-            stacks: Stacks::new(&env),
             input,
             output: Vec::new(),
-            workspace: 0,
+            registers: iter::repeat(0)
+                .take(usize::try_from(env.num_registers).unwrap())
+                .collect(),
+            stacks: Stacks::new(&env),
```
Comment (JRMurr) [inline api/src/lang/machine.rs]: still confused on this iter.repeat(0) stuff

File: api/src/lang/parse.rs [hunk 2]
```diff
@@ -1,101 +1,154 @@
...
 };
 
-fn parse_read(input: &str) -> IResult<&str, Instr> {
-    // tag_no_case returns a (str, str) tuple
-    // first element is whats left to parse
-    // second is what matched
-    let (input, _) = tag_no_case("Read")(input)?;
-    Ok((input, Instr::Read))
+fn arg_delim(input: &str) -> IResult<&str, char> {
+    one_of(" \t")(input)
```
Comment (JRMurr) [inline api/src/lang/parse.rs]: I dont think this will catch multiple spaces, if we want them to be able to do that

Comment (LucasPickering) [reply api/src/lang/parse.rs]: Yeah so I wanted to make it match multiple spaces without matching newlines and just didn't bother figuring out the repeating part, I figured I'd leave that to you. Are you cool with leaving it for now and you can change it in your PR?

File: api/src/lang/machine.rs (hunk 1 above)
Comment (LucasPickering) [reply api/src/lang/machine.rs]: Repeat just infinitely repeats whatever value you give it, so in this case it's repeating 0 once per register.

Comment (JRMurr) [reply api/src/lang/machine.rs]: o that makes more sense

File: api/src/lang/parse.rs (hunk 2 above)
Comment (JRMurr) [reply api/src/lang/parse.rs]: yea thats fine

**Pull request merged**